### DISCUSSION TITLE
Can Metals function without a workspace?

### DISCUSSION THREAD

**Discussion by oblitum** [scalameta/metals#3079] [General]
Hi.

Before tackling this question, I urge you first to read the following thread and associated discussion carefully, as I rather not replicate it again:

- https://github.com/scalameta/nvim-metals/pull/67

I'd just like to add upon that that, in the end, what's coming up of current situation is that conforming clients that send `null` when a workspace can't be inferred¹, will get the server running, but broken, from what I could gather, while other clients, like the referred one in the thread above, will implicit figure out a workspace to appease the server and not the protocol (notice though that for the latter, that could also be a client's issue, it could just not configure a workspace at all and warn/ask users about that).

You have been hit by the _nil workspace_ police 👮🏽‍♂️.

Disclaimer: I'm neither a Metals user or Scala programmer.

¹ : coc.nvim is one client known to do that, and that's _the_ reason it has solved issues with servers scanning the entire home folder because it used to occasionally send `~` as a workspace due to naïve reasoning, like picking a parent's file directory as workspace for example. It now instead follows the spec and sends `null` when it can't figure a workspace, and most servers are happy with that, by working in "no workspace"-mode. Knowing whether Metals **can not work without a workspace** will be useful so that coc.nvim should provide their example configuration with [requireRootPattern](https://github.com/neoclide/coc.nvim/blob/0feaa564cfeee696a2454b4664dde29d4675f77e/doc/coc.txt#L1048-L1049) set for this server in specific, so to avoid sending `null` for it, _and_ making it clear to users that a workspace folder is a precondition for this server to work (though the more conformant would be for servers to accept `null` and fail gracefully instead of expecting LSP clients being careful at sending `null` to specific servers).

**Comment by ckipp01** [comment]
To clarify a couple points and to continue the discussion from the other repo:

> ~will get the server blowing up~  will get the server running, but broken, from what I could gather

Metals won't blow up. If you open a standalone file and no workspace is sent in, you'll get a warning from Metals telling you that nothing is really going to work. The changes that made this happen are here: https://github.com/scalameta/metals/pull/2391

> will implicit figure out a workspace to appease the server and not the protocol

Again, this is an explicit decision made my `nvin-metals` to enable users to have support for standalone Scala files. This has never caused any issues.

> coc.nvim is one client known to do that, and that's the reason it has solved issues with servers scanning the entire home folder because it used to occasionally send ~ as a workspace due to naïve reasoning, like picking a parent's file directory as workspace for example. It now instead follows the spec and sends null when it can't figure a workspace, and most servers are happy with that, by working in "no workspace"-mode. Knowing whether Metals can not work without a workspace will be useful so that coc.nvim should provide their example configuration with requireRootPattern set for this server in specific, so to avoid sending null for it, and making it clear to users that a workspace folder is a precondition for this server to work (though the more conformant would be for servers to accept null and fail gracefully instead of expecting LSP clients being careful at sending null to specific servers).

On our side, we'll need to figure out a way to handle this in https://github.com/scalameta/coc-metals since this change will break support for standalone Scala files for `coc-metals` users. Heads up on this @jsatk.

So to answer your question, no, Metals will not function fully without a workspace. It will result in a `None` here, warn the user that there is no root, and then not bootstrap the majority of the functionality Metals has to offer.

https://github.com/scalameta/metals/blob/7d665e73aa5b8f55aba158f01c158a8d81c75152/metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala#L320-L321

**Reply by oblitum** [reply]
Please edit your answer as I've edited mine before you sent yours.

Edit: thanks for the edit.

**Reply by oblitum** [reply]
> Metals won't blow up.

Thanks for the answer's update <sub>(although, I'd prefer already outdated terms to be left behind as well)</sub>.

>  If you open a standalone file and no workspace is sent in, you'll get a warning from Metals telling you that nothing is really going to work. The changes that made this happen are here: #2391
> 
> > will implicit figure out a workspace to appease the server and not the protocol
> 
> Again, this is an explicit decision made my `nvin-metals` to enable users to have support for standalone Scala files.

I noticed that. That's why in my opinion this topic should have been _actually resolved_ in that thread, which got preemptively locked, so, there was no option but ask what's the server's stance here, as you have pointed me to do. Although it's confusing to me what does it mean in that "it's an explicit decision" but an "implicit behavior" (given that it's on its own, "out of the box", picking parent directories as workspace)? 

> So to answer your question, no, Metals will not function fully without a workspace. It will result in a `None` here, warn the user that there is no root, and then not bootstrap the majority of the functionality Metals has to offer.

That's correct, and even though it's not providing users with LSP features in such case, it's ultimately following the spec at that. So, in the end, there's nothing to blame the server's implementation here.

**Comment by tgodzik** [comment]
> You have been hit by the nil workspace police 👮🏽‍♂️.

Please refrain from using such a term or giving out warnings if you are not indeed a part of any kind of law enforcement agency. We try to keep the responses and comments positive and helpful to bolster a good community to interact with.

> Disclaimer: I'm neither a Metals user or Scala programmer.

If you are not using Metals, why do you need this change? I agree that we should allow to use Metals when no workspace was created, but at the current time we were not able to address the issue and this is not high on the priority list. We would welcome any help or contributions, but otherwise we can only move it to the feature requests repository.

**Reply by oblitum** [reply]
> > You have been hit by the nil workspace police 👮🏽‍♂️.
> 
> Please refrain from using such a term or giving out warnings if you are not indeed a part of any kind of law enforcement agency. We try to keep the responses and comments positive and helpful to bolster a good community to interact with.

OK. I'm sorry at that, it was just a joke 😅. Imagine if there was a real law enforcement agency for this? 😄 

> If you are not using Metals, why do you need this change?

I'm a long time coc.nvim user and occasional contributor in pointing such things and have been following the LSP spec and its issues for quite a long time, but at the end all I want is my editor working rock solid and this is a relevant, tricky, topic in specific, on which I've witnessed many kinds of issues, as can be inferred if you read up the discussion in depth. And I wish the same for other users as well, [that them don't get into the same traps](https://github.com/sumneko/lua-language-server/issues/480#issue-845528175). I could use Scala in the future too, so I rather have these issues figured out in the past.

> I agree that we should allow to use Metals when no workspace was created, but at the current time we were not able to address the issue and this is not high on the priority list. We would welcome any help or contributions, but otherwise we can only move it to the feature requests repository.

I don't see any problem with current Metals situation, [it's just fine](https://github.com/scalameta/metals/discussions/3079#discussioncomment-1255183). If it ever manages to provide LSP coding assistance without a workspace, then even better, but that's not a requirement (what _is_ required is that it should at least handle `null` gracefully, which it already does). I wish you the best at that.

**Reply by tgodzik** [reply]
Ok, so the situation in the vim plugin is that it could index the home directory? Is that it? We could check for the parent to make sure it's not home or alternatively open it in a temporary directory. Would that be satisfactory?

**Reply by oblitum** [reply]
> Ok, so the situation in the vim plugin is that it could index the home directory?  Is that it?

That's a case, but it's not _just_ that. I provide many illustrative examples at https://github.com/scalameta/nvim-metals/pull/67, please use these terms while searching there and you'll hit them: "/tmp", "read-only directory".

> We could check for the parent to make sure it's not home

Even though weird, an user might want that for some obscure reason, and it would be unable to because it's hard-coded, so, I don't think that's a very clean solution. Now, swapping clients at this topic, that's currently a problem on coc.nvim implementation side's too, it has a home check hard-coded, and sends `null` when home is resolved as a workspace root on client's side. I've already pointed them that it would be best instead to have their `ignoreRootPaths` with some explicitly documented defaults like `/`, `/tmp`, `~`, read-only dirs, etc and completely getting rid of internal hard-codes and concentrate their decision on `ignoreRootPaths`. That's something relevant to coc.nvim in specific, but clients can handle this in many ways possible, they could even prompt the user for a directory when it can't figure out a workspace root, they can just send `null` and present the negative (or positive) server's answer to users, etc.

> or alternatively open it in a temporary directory. Would that be satisfactory?

I'm not sure what that means exactly, whether that would be a client or server responsibility, but at that, if it's the case that when a client sends `null`, a server can figure itself with some `null` workspace specific behavior by leveraging `/tmp`, then that means starting to support LSP coding assistance without a workspace, on user's point of view. Such `/tmp`/`null` workspace would be a server's internal business.

**Reply by tgodzik** [reply]
In the end it should be solved internally in the server, but until then we will most likely leave the current behaviour of the plugins.

I do understand the concern here, but I think it's more of a benefit than it is danger.

**Reply by oblitum** [reply]
Okay.

**Reply by oblitum** [reply]
> In the end it should be solved internally in the server

@tgodzik do you think then it's worth opening an issue/feature-request for tracking that? If kept just as discussion it may just get lost or forgotten about with time. Feel free to do so as I'm not currently an user, so possibly not the best person to do that now. I'd subscribe to it though, to eventually learn how things have gone.

**Reply by tgodzik** [reply]
Raised it here: https://github.com/scalameta/metals-feature-requests/issues/231

It can be voted for and we usually try to pick things from the top of the list.

**Reply by oblitum** [reply]
@tgodzik thank you.
